INT-3945: Async Service Activator

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3945

garyrussell committed:
INT-3945: Async Service Activator

JIRA: https://jira.spring.io/browse/INT-3945

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/AbstractSimpleMessageHandlerFactoryBean.java:
```diff
@@ -122,6 +124,14 @@ public void setAdviceChain(List<Advice> adviceChain) {
+	/*
+	 * Supported at the super class level but only exposed on the service activator
+	 * namespace. It's not clear that other endpoints would benefit from async support.
+	 */
```
`@since 4.3`

garyrussell replied:
ok; but notice it's not javadoc - its doc for us

artembilan replied:
OK. That makes sense, too.
But I'm sure it can become a JavaDoc, since this method is `public`. 
Or even must!

With the growing community there is a big chance that someone will complain that we don't have a JavaDoc on this method :wink:

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/handler/ServiceActivatorDefaultFrameworkMethodTests.java:
```diff
@@ -221,4 +277,19 @@ public String processMessage(Message<?> message) {
 		}
 	}
 
+	private static class AsyncService {
```
We can add `RedundantModifier` to the `checkstyle-suppressions.xml` for the tests, but maybe it would be better to keep similar classes as `public`?
E.g. I fixed several of them recently because of automatic `fixModifiers`.

artembilan replied:
Ah! This one passes the check because it doesn't have an explicit `public` ctor like those :smile:

artembilan commented on src/reference/asciidoc/service-activator.adoc:
```diff
@@ -106,3 +106,20 @@ For simple scenarios your _Service Activators_ do not even have to reference a b
+This is particularly advantageous for long-running services using a `PollableChannel` because the poller thread is
+freed up to perform other services within the framework.
+
+If the service completes the future with an `Exception`, normal error processing will occur (an `ErrorMessage` is
```
I think it would be better to replace `(` to the `:`.
From other hand that leads me to the thought if we should use `errorChannel` bean instead of:
```
if (errorChannel == null) {
	logger.error("Async exception received and no 'errorChannel' header exists; cannot route "
								+ "exception to caller", result);
}
```
in the `AbstractMessageProducingHandler`.

garyrussell replied:
Yes, good idea :+1:

garyrussell committed:
Polishing

Send errors to the default `errorChannel` (if available) and no `errorChannel`
header present.

artembilan commented:
Not sure if related, but Travis barks a bit:
```
org.springframework.integration.handler.AsyncHandlerTests > testMessagingExceptionNoErrorChannel FAILED

    java.lang.AssertionError at AsyncHandlerTests.java:223
```

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java:
```diff
@@ -336,4 +316,41 @@ protected boolean shouldCopyRequestHeaders() {
+	}
+
+	protected void sendErrorMessage(final Message<?> requestMessage, Throwable ex) {
+		Object errorChannel = resolveErrorChannel(requestMessage.getHeaders());
```
The proper convention to declare used method after the user.
Minor, of course, but for "heads up!"  anyway.

artembilan commented on spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AsyncAmqpOutboundGateway.java:
```diff
@@ -127,30 +126,7 @@ public void onFailure(Throwable ex) {
-					}
-					logger.error("Failed to send async reply", exceptionToLog);
-				}
+				sendErrorMessage(this.requestMessage, exceptionToSend);
```
Since I don't see in the http://docs.spring.io/spring-integration/docs/4.3.0.BUILD-SNAPSHOT/reference/html/amqp.html#amqp-async-outbound-gateway any general error handling description, I think that should link to the `<<>>`, at least.
Although I see that `requires-reply` should be improved for the `errorChannel` bean anyway.
WDYT?

garyrussell committed:
Fix Test; Javadoc Polishing

garyrussell commented:
Pushed

garyrussell commented:
But I didn't address your latest comments yet

artembilan commented:
> But I didn't address your latest comments yet

No problem!
If you agree, I can address on merge.
Everything else LGTM here already

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/8db0ad3ae69dfa392b530dc7b0ead5e1398066ca
